infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/StartSynchronizer.java

```diff
@@ -0,0 +1,315 @@
+ *     startSynchronizer.setStartThread();
+ *     try {
+ *       startMachines();
+ *       startSynchronizer.checkFailure();
+ *
+ *       // ...
+ *
+ *       waitMachines()
+ *         .then(startSynchronizer.checkFailure())
+ *         .then(bootstrapAsync())
+ *         .then(startSynchronizer.checkFailure())
+ *         .then(checkServers)
+ *         // ...
+ *
+ *       startSynchronizer.complete();
+ *     } catch (Exception ex) {
+ *       startSynchronizer.completeExceptionally(ex);
+ *       throw ex;
+ *     }
+ *   }
+ * ...
+ * </pre>
+ *
+ * <p>At the same time stopping might look like:
+ *
+ * <pre>
+ * ...
+ *   public void stopRuntime() {
+ *     if (startSynchronizer.interrupt()) {
+ *       try {
+ *         if (startSynchronizer.awaitInterruption(30)) {
+ *           // runtime is interrupted correctly
+ *         } else {
+ *           // runtime is not interrupted correctly in time
+ *           // need to forcibly stop it and clean up used resources
+ *         }
+ *       } catch (RuntimeStartInterruptedException ex) {
+ *         // normal stop
+ *       } catch (InterruptedException ex) {
+ *         Thread.currentThread().interrupt();
+ *         ...
+ *       }
+ *     }
+ *   }
+ * ...
+ * </pre>
+ *
+ * <p>Note that class is designed to work in clustered mode. For this purpose it listens to {@link
+ * KubernetesRuntimeStoppedEvent} and {@link KubernetesRuntimeStoppingEvent} events. So, if Che
+ * Server is run in clustered mode then the described events must be published via {@link
+ * EventService} otherwise start interruption won't work correctly.
+ *
+ * @author Sergii Leshchenko
+ */
+public class StartSynchronizer {
+
+  private final RuntimeIdentity runtimeId;
+  private final EventService eventService;
+
+  // future that holds error that occurs during runtime start
+  // failure must be completed with null value when start is finished without any exception
+  private final CompletableFuture<Void> startFailure;
+
+  // latch that indicates whether start is completed or not
+  private CountDownLatch completionLatch;
+
+  // thread which is performing start
+  // it may be nullable when start is performing on another Che Server instance
+  private Thread startThread;
+
+  // flag that indicates whether start is in progress or not
+  private boolean isStarting;
+
+  private final RuntimeStartInterrupter runtimeStartInterrupter;
+  private final RuntimeStopWatcher runtimeStopWatcher;
+
+  public StartSynchronizer(RuntimeIdentity runtimeId, EventService eventService) {
+    this.eventService = eventService;
+    this.startFailure = new CompletableFuture<>();
+    this.completionLatch = new CountDownLatch(0);
+    this.runtimeId = runtimeId;
+    this.runtimeStartInterrupter = new RuntimeStartInterrupter();
+    this.runtimeStopWatcher = new RuntimeStopWatcher();
+    this.isStarting = false;
+  }
+
+  /** Registers a runtime start. */
+  public synchronized void start() {
+    if (!isStarting) {
```

Comment:
Can you elaborate what will happen if 2 servers will try to simultaneously start a runtime? Is there a protection from that on this or some another level?